source/extensions/filters/http/aws_lambda/sts_credentials_provider.cc

```diff
@@ -57,28 +59,59 @@ class StsCredentialsProviderImpl : public StsCredentialsProvider,
 };
 
 StsCredentialsProviderImpl::StsCredentialsProviderImpl(
-    const envoy::config::filter::http::aws_lambda::v2::
-        AWSLambdaConfig_ServiceAccountCredentials &config,
-    Api::Api &api, Upstream::ClusterManager &cm,
-    StsConnectionPoolFactoryPtr conn_pool_factory, std::string_view web_token,
-    std::string_view role_arn)
-    : api_(api), cm_(cm), config_(config), default_role_arn_(role_arn),
-      conn_pool_factory_(std::move(conn_pool_factory)), web_token_(web_token) {
-  // file_watcher_(dispatcher.createFilesystemWatcher()) {
+  const envoy::config::filter::http::aws_lambda::v2::
+      AWSLambdaConfig_ServiceAccountCredentials &config,
+  Api::Api &api, Upstream::ClusterManager &cm,
+  StsConnectionPoolFactoryPtr conn_pool_factory, std::string_view web_token,
+  std::string_view role_arn)
+  : api_(api), cm_(cm), config_(config), default_role_arn_(role_arn),
+    conn_pool_factory_(std::move(conn_pool_factory)), web_token_(web_token) {
+  
 
   uri_.set_cluster(config_.cluster());
   uri_.set_uri(config_.uri());
   // TODO: Figure out how to get this to compile, timeout is not all that
   // important right now uri_.set_allocated_timeout(config_.mutable_timeout())
+  // Consider if this is still reasonable. We have a timeout configuration for
+  // base webtoken handled in a filewatch in the config init.
+  // We do want to decide about whether it matters to pass for chained.
+  // Though chained can only have expiry of 15m to 60m and defaults to 60m.
 }
 
 void StsCredentialsProviderImpl::setWebToken(std::string_view web_token) {
   web_token_ = web_token;
 }
 
 void StsCredentialsProviderImpl::onResult(
-    std::shared_ptr<const StsCredentials> result, std::string_view role_arn) {
+    std::shared_ptr<const StsCredentials> result, std::string role_arn,
+    std::list<std::string> chained_requests) {
   credentials_cache_.emplace(role_arn, result);
+
+  // kick off any waiting chained assumption roles relying on this credential
+  while( !chained_requests.empty()){
+    auto chained_role = chained_requests.back();
+    ENVOY_LOG(trace, "calling sts chained for {}", chained_role);
+    auto conn_pool = connection_pools_.find(chained_role);
+    if (conn_pool != connection_pools_.end()) {
+      conn_pool->second->init(uri_, "", result);
+    }
+    chained_requests.pop_back();
+  }
+}
+
+void StsCredentialsProviderImpl::onFailure(CredentialsFailureStatus status,
+                                      std::list<std::string> chained_requests) {
```

Comment:
do we need to modify the original list? if so, we can use reference (&). If not, we can use move (&&).
same for onResult

Reply:
We dont need to modify the original list but it is the cleanest way. 
I have updated accordingly